lib/spack/spack/util/web.py

```diff
@@ -212,7 +212,11 @@ def url_exists(url):
         s3 = s3_util.create_s3_session(url)
         from botocore.exceptions import ClientError
         try:
-            s3.get_object(Bucket=url.netloc, Key=url.path)
+            key = url.path
+            if key.startswith('/'):
+                key = key[1:]
```

Comment:
You also need to check if the path is only `/`, otherwise checks for just a bucket will fail (e.g. `s3://my-bucket` will fail but `s3://my-bucket/my-file` will not).

Additionally, I'd recommend using `key.lstrip('/')` instead of `key[1:]` so that if there are multiple leading slashes they are removed appropriately.

Reply:
The case of `scheme == 's3' and path == '/'` is curious.  We do need this to be a valid case for listing (show me everything in the top level of my bucket), but I probably need to update `url_exists()` to handle this specific case (I'd interpret it as asking "does the bucket exist, at all?"), and I don't think this case even makes sense for fetching.  In that case I think Spack should just raise an exception without even trying.

Good point on the use of `lstrip()`!

Reply:
Actually, I may have just talked myself out of my interpretation for `url_exists()`, since I imagine that for most purposes, a generic resources "exists" if and only if I can "fetch" it.  I even rely on this assumption as a fallback (just try to fetch, and assume that any successful result implies existence).

`url_exists()` for the above case should probably just return `False`, since fetching a bucket is undefined.

(I realize this isn't directly related to getting your use case working, again, but I figure this is probably a good spot to sort this out.)

Reply:
There's another method you could use called `head_bucket`, but it does technically require the `ListBucket` permission, but that's probably not too much to ask. This method `url_exists()` should definitely be able to check if a bucket exists, but you could modify the logic to call `head_bucket` if the `url.path` is just `/` instead of trying to fetch it. I think the fetch is lazy when it's the whole bucket anyway.

Reply:
Just tested this locally and it seems to work:
```python
if url.path == '/':
  s3.head_bucket(Bucket=url.netloc)
else:
  s3.get_object(Bucket=url.netloc, Key=url.path.lstrip('/'))
```

Reply:
I think I'll need to do something like what you're suggesting, after all.  It's occurred to me that fetchable implies exists, but not vice-a-versa.  File system directories are another example and I'm pretty sure we rely on `url_exists()` to return `True` in this case.

So exists implies *either* fetchable or ... "listable"?

Man, what a rabbit hole.  Thanks for humoring me! 😃